cpp/include/cudf/rolling.hpp

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/types.hpp>
+
+#include <memory>
+
+namespace cudf {
+namespace experimental {
+
+/** 
+ * @brief Rolling window aggregation operations
+ */
+enum class rolling_operator {
+  SUM,       ///< Computes the sum of all values in the window
+  MIN,       ///< Computes minimum value in the window
+  MAX,       ///< Computes maximum value in the window
+  MEAN,       ///< Computes arithmetic mean of all values in the window
+  COUNT,     ///< Computes the number of values in the window
+  NUMBA_UDF, ///< A user-defined aggregation operation defined in PTX code generated by `numba`
+  CUDA_UDF,  ///< A generic aggregation operation defined in CUDA code
+  N_OPS,     ///< ALL NEW OPERATIONS SHOULD BE ADDED ABOVE THIS LINE
+};
+
+/**
+ * @brief  Applies a fixed-size rolling window function to the values in a column.
+ *
+ * This function aggregates values in a window around each element i of the input column, and
+ * invalidates the bit mask for element i if there are not enough observations. The window size is
+ * static (the same for each element). This matches Pandas' API for DataFrame.rolling with a few
+ * notable differences:
+ * - instead of the center flag it uses the forward window size to allow for more flexible windows.
+ *   The total window size = window + forward_window. Element i uses elements
+ *   [i-window+1, i+forward_window] to do the window computation.
+ * - instead of storing NA/NaN for output rows that do not meet the minimum number of observations
+ *   this function updates the valid bitmask of the column to indicate which elements are valid.
+ * 
+ * The returned column for `op == COUNT` always has INT32 type.
+ *
+ * @param[in] input_col The input column
+ * @param[in] window The static rolling window size.
```

Comment:
Nope, pandas doesn't have a two-part  window. This naming comes from @nsakharnykh ...

Reply:
See above my suggestion to rename this to preceding and following window sizes, that would address both concerns about inclusiveness/exclusiveness and the direction. We would diverge from gQuant's (exclusive `window`, inclusive `forward_window`) and Panda's API (exclusive `window`, `centered` parameter), but will be more compatible with SQL.

Reply:
BTW, it was more than just renaming. I also had to change the behavior to compute the window as [i - preceding_window, i + following_window] instead of [i - window + 1, i + forward_window]. But it all works. Python will just have to change the bindings to account for this. @kkraus14 @shwina FYI.